Environment API - Cross env poisoning via hooks

### Describe the bug

When defining 2 or more environments (using [Environment API](https://vite.dev/guide/api-environment)) in the Vite config, mutations made by a plugin's [`configEnvironment`](https://github.com/vitejs/vite/blob/44c42b976/packages/vite/src/node/config.ts#L2638) hook in one environment can leak into other environments.

This happens because [`resolveConfig`](https://github.com/vitejs/vite/blob/44c42b976/packages/vite/src/node/config.ts#L1554-L1565) constructs a single [`defaultNonClientEnvironmentOptions`](https://github.com/vitejs/vite/blob/44c42b976/packages/vite/src/node/config.ts#L1554) object and reuses it for all non-client environments in [the merge loop](https://github.com/vitejs/vite/blob/44c42b976/packages/vite/src/node/config.ts#L1567-L1574). Since [`mergeConfig`](https://github.com/vitejs/vite/blob/44c42b976/packages/vite/src/node/utils.ts#L1337) does a shallow spread, nested objects like `resolve` retain their shared reference. When a plugin then mutates `config.resolve` for one environment, all other non-client environments see the same change.

In practice, this surfaces with Vitest when using 2+ environments (from Environment API) and [custom environment (from Vitest)](https://vitest.dev/guide/environment.html#custom-environment) - Vitest's [`configEnvironment`](https://github.com/vitest-dev/vitest/blob/21e66ff6367f87ef91a70e4c44315a89644d35c5/packages/vitest/src/node/plugins/runnerTransform.ts#L78-L116) hook sets [`resolve.noExternal = true`](https://github.com/vitest-dev/vitest/blob/21e66ff6367f87ef91a70e4c44315a89644d35c5/packages/vitest/src/node/plugins/runnerTransform.ts#L109) per environment, which propagates to all non-client environments, affecting [this condition for the next configEnvironment call](https://github.com/vitest-dev/vitest/blob/21e66ff6367f87ef91a70e4c44315a89644d35c5/packages/vitest/src/node/plugins/runnerTransform.ts#L86-L89), causing CJS dependencies (like jsdom) to crash with `ReferenceError: require is not defined`. Take this with a grain of salt as this is only my theory and I don't have full understanding of Vite/Vitest.

If you name an environment `client`, it is unaffected because it gets a separately constructed [`defaultClientEnvironmentOptions`](https://github.com/vitejs/vite/blob/44c42b976/packages/vite/src/node/config.ts#L1549). A single non-client environment also works fine since there's nothing to leak to.

### Reproduction

https://github.com/Filipoliko/vitest-environment-bug

### Steps to reproduce

Run `npm ci` followed by `npm test` in the reproduction repository. Check out the README file in the project for more info.

Check out also PR #22250 for a unit test in this repository that also showcases this issue.

### System Info

```shell
System:
    OS: macOS 26.4.1
    CPU: (10) arm64 Apple M1 Pro
    Memory: 147.92 MB / 16.00 GB
    Shell: 5.2.37 - /opt/homebrew/bin/bash
  Binaries:
    Node: 24.14.1 - /Users/filip.satek/.nvm/versions/node/v24.14.1/bin/node
    npm: 11.11.0 - /Users/filip.satek/.nvm/versions/node/v24.14.1/bin/npm
    pnpm: 10.33.0 - /Users/filip.satek/.nvm/versions/node/v24.14.1/bin/pnpm
    bun: 1.2.12 - /opt/homebrew/bin/bun
  Browsers:
    Chrome: 147.0.7727.57
    Edge: 113.0.1774.57
    Firefox: 149.0.2
    Safari: 26.4
```

### Used Package Manager

npm

### Logs

```
$ npm test

> vitest-environment@1.0.0 test
> vitest run


 RUN  v4.1.4 /Users/filip.satek/git/vitest-environment

⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯ Unhandled Errors ⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯

Vitest caught 1 unhandled error during the test run.
This might cause false positive tests. Resolve unhandled errors to make sure your tests are not affected.

⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯ Unhandled Error ⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯
Error: [vitest-pool]: Failed to start forks worker for test files /Users/filip.satek/git/vitest-environment/test.spec.ts.
 ❯ node_modules/vitest/dist/chunks/cli-api.lDy4N9kC.js:3448:94
 ❯ processTicksAndRejections node:internal/process/task_queues:104:5
 ❯ Pool.schedule node_modules/vitest/dist/chunks/cli-api.lDy4N9kC.js:3448:5

Caused by: ReferenceError: require is not defined
 ❯ eval node_modules/vite/dist/node/module-runner.js:992:9
 ❯ ESModulesEvaluator.runInlinedModule node_modules/vite/dist/node/module-runner.js:992:161
 ❯ ModuleRunner.directRequest node_modules/vite/dist/node/module-runner.js:1247:80
 ❯ processTicksAndRejections node:internal/process/task_queues:104:5
 ❯ ModuleRunner.cachedRequest node_modules/vite/dist/node/module-runner.js:1154:73

⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯


 Test Files  no tests
      Tests  no tests
     Errors  1 error
   Start at  12:31:44
   Duration  238ms (transform 117ms, setup 0ms, import 0ms, tests 0ms, environment 0ms)
```

### Validations

- [x] Follow our [Code of Conduct](https://github.com/vitejs/vite/blob/main/CODE_OF_CONDUCT.md)
- [x] Read the [Contributing Guidelines](https://github.com/vitejs/vite/blob/main/CONTRIBUTING.md).
- [x] Read the [docs](https://vite.dev/guide).
- [x] Check that there isn't [already an issue](https://github.com/vitejs/vite/issues) that reports the same bug to avoid creating a duplicate.
- [x] Make sure this is a Vite issue and not a framework-specific issue. For example, if it's a Vue SFC related bug, it should likely be reported to [vuejs/core](https://github.com/vuejs/core) instead.
- [x] Check that this is a concrete bug. For Q&A open a [GitHub Discussion](https://github.com/vitejs/vite/discussions) or join our [Discord Chat Server](https://chat.vite.dev/).
- [x] The provided reproduction is a [minimal reproducible example](https://stackoverflow.com/help/minimal-reproducible-example) of the bug.

--- a/packages/vite/src/node/__tests__/config.spec.ts
+++ b/packages/vite/src/node/__tests__/config.spec.ts
@@ -1116,6 +1116,25 @@ test('configEnvironment mutation does not leak between environments', async () =
   expect(resolved.environments.custom2.resolve.noExternal).not.toBe(true)
 })
 
+test('build and environments.client.build has the same reference', async () => {
+  const nameCache = {}
+  const resolved = await resolveConfig(
+    {
+      build: {
+        terserOptions: {
+          nameCache,
+        },
+      },
+    },
+    'serve',
+  )
+
+  expect(resolved.build.terserOptions.nameCache).toBe(nameCache)
+  expect(resolved.environments.client.build.terserOptions.nameCache).toBe(
+    resolved.build.terserOptions.nameCache,
+  )
+})
+
 test('preTransformRequests', async () => {
   async function testConfig(inlineConfig: InlineConfig) {
     return Object.fromEntries(

--- a/packages/vite/src/node/__tests__/config.spec.ts
+++ b/packages/vite/src/node/__tests__/config.spec.ts
@@ -1090,6 +1090,32 @@ test('config compat 3', async () => {
   `)
 })
 
+test('configEnvironment mutation does not leak between environments', async () => {
+  const resolved = await resolveConfig(
+    {
+      environments: {
+        custom1: {},
+        custom2: {},
+      },
+      plugins: [
+        {
+          name: 'test-mutate-env',
+          configEnvironment(name, config) {
+            if (name === 'custom1') {
+              config.resolve ??= {}
+              config.resolve.noExternal = true
+            }
+          },
+        },
+      ],
+    },
+    'serve',
+  )
+
+  expect(resolved.environments.custom1.resolve.noExternal).toBe(true)
+  expect(resolved.environments.custom2.resolve.noExternal).not.toBe(true)
+})
+
 test('preTransformRequests', async () => {
   async function testConfig(inlineConfig: InlineConfig) {
     return Object.fromEntries(

--- a/packages/vite/src/node/config.ts
+++ b/packages/vite/src/node/config.ts
@@ -80,6 +80,7 @@ import {
   asyncFlatten,
   createDebugger,
   createFilter,
+  deepClone,
   hasBothRollupOptionsAndRolldownOptions,
   isExternalUrl,
   isFilePathESM,
@@ -1568,7 +1569,9 @@ export async function resolveConfig(
     config.environments[name] = mergeConfig(
       name === 'client'
         ? defaultClientEnvironmentOptions
-        : defaultNonClientEnvironmentOptions,
+        : (deepClone(
+            defaultNonClientEnvironmentOptions as object,
+          ) as UserConfig),
       config.environments[name],
     )
   }

